TestFileStoreAgeLimitRecovery failed

## Defect

When running nats server unit tests I ran into the failure below.
I reran unit tests a few more times and did not run into this again.

```
=== RUN   TestFileStoreAgeLimitRecovery
--- FAIL: TestFileStoreAgeLimitRecovery (0.04s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x171ab8e]

goroutine 8003 [running]:
testing.tRunner.func1(0xc0005ca100)
/usr/local/go/src/testing/testing.go:874 +0x69f
panic(0x1bb5ee0, 0x23c8190)
/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/nats-io/nats-server/v2/server.(*fileStore).deleteMsgFromBlock(0xc0005cc540, 0xc0002c1a40, 0x62, 0xc0008b2f60, 0x0)
/Users/matthiashanel/repos/nats-server/server/filestore.go:781 +0x11e
github.com/nats-io/nats-server/v2/server.(*fileStore).removeMsg(0xc0005cc540, 0x62, 0x0, 0x0, 0x62, 0x0)
/Users/matthiashanel/repos/nats-server/server/filestore.go:717 +0x100
github.com/nats-io/nats-server/v2/server.(*fileStore).deleteFirstMsg(0xc0005cc540, 0x62, 0xc0008b2ea0, 0x0)
/Users/matthiashanel/repos/nats-server/server/filestore.go:675 +0x85
github.com/nats-io/nats-server/v2/server.(*fileStore).expireMsgs(0xc0005cc540)
/Users/matthiashanel/repos/nats-server/server/filestore.go:897 +0x131
github.com/nats-io/nats-server/v2/server.(*fileStore).expireMsgsLocked(0xc0005cc540)
/Users/matthiashanel/repos/nats-server/server/filestore.go:885 +0x49
github.com/nats-io/nats-server/v2/server.(*fileStore).recoverMsgs(0xc0005cc540, 0x0, 0x0)
/Users/matthiashanel/repos/nats-server/server/filestore.go:467 +0x61c
github.com/nats-io/nats-server/v2/server.newFileStoreWithCreated(0xc0003c3590, 0x43, 0x4000000, 0x12a05f200, 0x2540be400, 0x1c7a4b4, 0x3, 0x0, 0x0, 0x0, ...)
/Users/matthiashanel/repos/nats-server/server/filestore.go:239 +0xa2c
github.com/nats-io/nats-server/v2/server.newFileStore(0xc0003c3590, 0x43, 0x0, 0x0, 0x0, 0x1c7a4b4, 0x3, 0x0, 0x0, 0x0, ...)
/Users/matthiashanel/repos/nats-server/server/filestore.go:175 +0x129
github.com/nats-io/nats-server/v2/server.TestFileStoreAgeLimitRecovery(0xc0005ca100)
/Users/matthiashanel/repos/nats-server/server/filestore_test.go:628 +0x582
testing.tRunner(0xc0005ca100, 0x1cd4e38)
/usr/local/go/src/testing/testing.go:909 +0x19a
created by testing.(*T).Run
/usr/local/go/src/testing/testing.go:960 +0x652
FAIL github.com/nats-io/nats-server/v2/server 35.451s
```

Ran the unit test to test my filter change. This is the last commit in main branch
```
commit c8b4b2efa39b38ac2a5a3dfd7b30257189109e5c (origin/master, origin/HEAD, master)
Date:   Mon Jun 15 10:39:07 2020 -0700
```

--- a/server/filestore.go
+++ b/server/filestore.go
@@ -65,22 +65,23 @@ type FileConsumerInfo struct {
 }
 
 type fileStore struct {
-	mu      sync.RWMutex
-	state   StreamState
-	scb     func(int64)
-	ageChk  *time.Timer
-	syncTmr *time.Timer
-	cfg     FileStreamInfo
-	fcfg    FileStoreConfig
-	lmb     *msgBlock
-	blks    []*msgBlock
-	hh      hash.Hash64
-	wmb     *bytes.Buffer
-	fch     chan struct{}
-	qch     chan struct{}
-	cfs     []*consumerFileStore
-	closed  bool
-	sips    int
+	mu       sync.RWMutex
+	state    StreamState
+	scb      func(int64)
+	ageChk   *time.Timer
+	syncTmr  *time.Timer
+	cfg      FileStreamInfo
+	fcfg     FileStoreConfig
+	lmb      *msgBlock
+	blks     []*msgBlock
+	hh       hash.Hash64
+	wmb      *bytes.Buffer
+	fch      chan struct{}
+	qch      chan struct{}
+	cfs      []*consumerFileStore
+	closed   bool
+	expiring bool
+	sips     int
 }
 
 // Represents a message store block and its data.
@@ -714,7 +715,9 @@ func (fs *fileStore) removeMsg(seq uint64, secure bool) (bool, error) {
 	sm, _ := mb.fetchMsg(seq)
 	// We have the message here, so we can delete it.
 	if sm != nil {
-		fs.deleteMsgFromBlock(mb, seq, sm, secure)
+		if err := fs.deleteMsgFromBlock(mb, seq, sm, secure); err != nil {
+			return false, err
+		}
 	}
 	return sm != nil, nil
 }
@@ -770,7 +773,7 @@ func (mb *msgBlock) selectNextFirst() {
 	}
 }
 
-func (fs *fileStore) deleteMsgFromBlock(mb *msgBlock, seq uint64, sm *fileStoredMsg, secure bool) {
+func (fs *fileStore) deleteMsgFromBlock(mb *msgBlock, seq uint64, sm *fileStoredMsg, secure bool) error {
 	// Update global accounting.
 	msz := fileStoreMsgSize(sm.subj, sm.hdr, sm.msg)
 
@@ -778,17 +781,34 @@ func (fs *fileStore) deleteMsgFromBlock(mb *msgBlock, seq uint64, sm *fileStored
 	mb.mu.Lock()
 
 	// Make sure the seq still exists.
+	if mb.cache == nil {
+		// Rare condition but can happen. We will load with lock held at this point.
+		buf, err := ioutil.ReadFile(mb.mfn)
+		if err == nil {
+			err = mb.indexCacheBuf(buf)
+		}
+		if err != nil {
+			mb.mu.Unlock()
+			fs.mu.Unlock()
+			return err
+		}
+		if len(buf) > 0 {
+			mb.cloads++
+			mb.startCacheExpireTimer()
+		}
+	}
+
 	if seq < mb.cache.fseq || (seq-mb.cache.fseq) >= uint64(len(mb.cache.idx)) {
 		mb.mu.Unlock()
 		fs.mu.Unlock()
-		return
+		return nil
 	}
 	// Now check dmap if it is there.
 	if mb.dmap != nil {
 		if _, ok := mb.dmap[seq]; ok {
 			mb.mu.Unlock()
 			fs.mu.Unlock()
-			return
+			return nil
 		}
 	}
 
@@ -846,6 +866,8 @@ func (fs *fileStore) deleteMsgFromBlock(mb *msgBlock, seq uint64, sm *fileStored
 		delta := int64(msz)
 		fs.scb(-delta)
 	}
+
+	return nil
 }
 
 // Lock should be held.
@@ -888,6 +910,21 @@ func (fs *fileStore) expireMsgsLocked() {
 
 // Will expire msgs that are too old.
 func (fs *fileStore) expireMsgs() {
+	// Make sure this is only running one at a time.
+	fs.mu.Lock()
+	if fs.expiring {
+		fs.mu.Unlock()
+		return
+	}
+	fs.expiring = true
+	fs.mu.Unlock()
+
+	defer func() {
+		fs.mu.Lock()
+		fs.expiring = false
+		fs.mu.Unlock()
+	}()
+
 	now := time.Now().UnixNano()
 	minAge := now - int64(fs.cfg.MaxAge)
 
@@ -1361,7 +1398,13 @@ func (mb *msgBlock) indexCacheBuf(buf []byte) error {
 func (mb *msgBlock) loadMsgs() error {
 	mb.mu.RLock()
 	mfn := mb.mfn
+	hasCache := mb.cache != nil
 	mb.mu.RUnlock()
+
+	// Someone else may have filled this in by the time we get here.
+	if hasCache {
+		return nil
+	}
 
 	// Load in the whole block.
 	buf, err := ioutil.ReadFile(mfn)

--- a/server/filestore_test.go
+++ b/server/filestore_test.go
@@ -607,7 +607,10 @@ func TestFileStoreAgeLimitRecovery(t *testing.T) {
 	os.MkdirAll(storeDir, 0755)
 	defer os.RemoveAll(storeDir)
 
-	fs, err := newFileStore(FileStoreConfig{StoreDir: storeDir}, StreamConfig{Name: "zzz", Storage: FileStorage, MaxAge: maxAge})
+	fs, err := newFileStore(
+		FileStoreConfig{StoreDir: storeDir, ReadCacheExpire: 1 * time.Millisecond},
+		StreamConfig{Name: "zzz", Storage: FileStorage, MaxAge: maxAge},
+	)
 	if err != nil {
 		t.Fatalf("Unexpected error: %v", err)
 	}
@@ -615,7 +618,7 @@ func TestFileStoreAgeLimitRecovery(t *testing.T) {
 
 	// Store some messages. Does not really matter how many.
 	subj, msg := "foo", []byte("Hello World")
-	toStore := 100
+	toStore := 1000
 	for i := 0; i < toStore; i++ {
 		fs.StoreMsg(subj, nil, msg)
 	}

--- a/server/memstore.go
+++ b/server/memstore.go
@@ -215,11 +215,17 @@ func (ms *memStore) expireMsgs() {
 			ms.deleteFirstMsgOrPanic()
 		} else {
 			if !ok {
-				ms.ageChk.Stop()
-				ms.ageChk = nil
+				if ms.ageChk != nil {
+					ms.ageChk.Stop()
+					ms.ageChk = nil
+				}
 			} else {
 				fireIn := time.Duration(sm.ts-now) + ms.cfg.MaxAge
-				ms.ageChk.Reset(fireIn)
+				if ms.ageChk != nil {
+					ms.ageChk.Reset(fireIn)
+				} else {
+					ms.ageChk = time.AfterFunc(fireIn, ms.expireMsgs)
+				}
 			}
 			return
 		}

